### PR TITLE
PIN-1377 Create clone endpoint

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala [hunk 1]
```diff
@@ -431,6 +432,62 @@ final case class PurposeApiServiceImpl(
     }
   }
 
+  override def clonePurpose(purposeId: String)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerPurpose: ToEntityMarshaller[Purpose],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem]
+  ): Route =
+    authorize(ADMIN_ROLE, API_ROLE, SECURITY_ROLE) {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]: Write operations on purposes can be done only by admins
```suggestion
    authorize(ADMIN_ROLE) {
```

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala [hunk 2]
```diff
@@ -431,6 +432,62 @@ final case class PurposeApiServiceImpl(
...
+
+        val ordering: Ordering[OffsetDateTime] = Ordering(Ordering.by[OffsetDateTime, Long](_.toEpochSecond).reverse)
+
+        val latestNoWaiting: Option[Int] = versions
+          .filterNot(_.state == PurposeManagementDependency.PurposeVersionState.WAITING_FOR_APPROVAL)
+          .sortBy(_.createdAt)(ordering)
+          .map(_.dailyCalls)
+          .headOption
+
+        val latestAll: Option[Int] = versions.sortBy(s => s.createdAt)(ordering).map(_.dailyCalls).headOption
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]:
```suggestion
        val latestAll: Option[Int] = versions.sortBy(_.createdAt)(ordering).map(_.dailyCalls).headOption
```

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala [hunk 3]
```diff
@@ -431,6 +432,62 @@ final case class PurposeApiServiceImpl(
     }
   }
 
+  override def clonePurpose(purposeId: String)(implicit
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala:433]: this endpoint needs to be tested with the `authz` test

File: src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala [hunk 4]
```diff
@@ -32,6 +32,414 @@ class PurposeApiServiceSpec extends AnyWordSpecLike with SpecHelper with Scalate
...
+            createdAt = OffsetDateTime.of(2022, 11, 30, 11, 22, 33, 44, ZoneOffset.UTC),
+            updatedAt = None,
+            firstActivationAt = None,
+            expectedApprovalDate = None,
+            dailyCalls = 50,
+            riskAnalysis = None
+          ),
+          PurposeManagementDependency.PurposeVersion(
+            id = UUID.randomUUID(),
+            state = PurposeManagementDependency.PurposeVersionState.SUSPENDED,
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala]: Just to have a more realistic case (only one version in a purpose can be ACTIVE or SUSPENDED in a given time)
```suggestion
            state = PurposeManagementDependency.PurposeVersionState.ARCHIVED,
```

File: src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala [hunk 5]
```diff
@@ -32,6 +32,414 @@ class PurposeApiServiceSpec extends AnyWordSpecLike with SpecHelper with Scalate
 
   import PurposeApiMarshallerImpl._
 
+  "Purpose cloning" should {
+    "succeed" in {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala]:
```suggestion
    "succeed when there is only a WAITING FOR APPROVAL version" in {
```

File: src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala [hunk 6]
```diff
@@ -32,6 +32,414 @@ class PurposeApiServiceSpec extends AnyWordSpecLike with SpecHelper with Scalate
...
+        .createPurposeVersion(_: UUID, _: PurposeManagementDependency.PurposeVersionSeed)(_: Seq[(String, String)]))
+        .expects(purposeCloned.id, *, context)
+        .once()
+        .returns(Future.successful(SpecData.purposeVersion))
+
+      Get() ~> service.clonePurpose(purposeId.toString) ~> check {
+        status shouldEqual StatusCodes.OK
+      }
+    }
+    "succeed with multiple versions latest is WAITING FOR APPROVAL" in {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala]:
```suggestion

    "succeed with multiple versions when latest is WAITING FOR APPROVAL" in {
```

File: src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala [hunk 7]
```diff
@@ -32,6 +32,414 @@ class PurposeApiServiceSpec extends AnyWordSpecLike with SpecHelper with Scalate
...
+        .createPurposeVersion(_: UUID, _: PurposeManagementDependency.PurposeVersionSeed)(_: Seq[(String, String)]))
+        .expects(purposeCloned.id, PurposeManagementDependency.PurposeVersionSeed(50, None), context)
+        .once()
+        .returns(Future.successful(SpecData.purposeVersion))
+
+      Get() ~> service.clonePurpose(purposeId.toString) ~> check {
+        status shouldEqual StatusCodes.OK
+      }
+    }
+    "succeed with multiple versions no WAITING FOR APPROVAL" in {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala]:
```suggestion

    "succeed with multiple versions when there is no WAITING FOR APPROVAL" in {
```

File: src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala [hunk 8]
```diff
@@ -32,6 +32,414 @@ class PurposeApiServiceSpec extends AnyWordSpecLike with SpecHelper with Scalate
...
+        .createPurposeVersion(_: UUID, _: PurposeManagementDependency.PurposeVersionSeed)(_: Seq[(String, String)]))
+        .expects(purposeCloned.id, PurposeManagementDependency.PurposeVersionSeed(100, None), context)
+        .once()
+        .returns(Future.successful(SpecData.purposeVersion))
+
+      Get() ~> service.clonePurpose(purposeId.toString) ~> check {
+        status shouldEqual StatusCodes.OK
+      }
+    }
+    "succeed with check of new purpose title" in {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala:264]:
```suggestion

    "succeed with check of new purpose title" in {
```

File: src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala [hunk 9]
```diff
@@ -32,6 +32,414 @@ class PurposeApiServiceSpec extends AnyWordSpecLike with SpecHelper with Scalate
...
+        .createPurposeVersion(_: UUID, _: PurposeManagementDependency.PurposeVersionSeed)(_: Seq[(String, String)]))
+        .expects(purposeCloned.id, PurposeManagementDependency.PurposeVersionSeed(500, None), context)
+        .once()
+        .returns(Future.successful(SpecData.purposeVersion))
+
+      Get() ~> service.clonePurpose(purposeId.toString) ~> check {
+        status shouldEqual StatusCodes.OK
+      }
+    }
+    "fail if Purpose does not exist" in {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/purposeprocess/PurposeApiServiceSpec.scala:342]:
```suggestion

    "fail if Purpose does not exist" in {
```

**Review by galales — CHANGES_REQUESTED** [review]

**Review by TonioGela — APPROVED** [review]

**Review by galales — APPROVED** [review]

**Pull request merged**